File: modules/core/src/main/java/de/deutschebahn/bahnhoflive/view/SelectableItemViewHolder.java
/*
 * SPDX-FileCopyrightText: 2020 DB Station&Service AG <bahnhoflive-opensource@deutschebahn.com>
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package de.deutschebahn.bahnhoflive.view;

import android.view.View;
import android.view.ViewGroup;

import de.deutschebahn.bahnhoflive.R;
import de.deutschebahn.bahnhoflive.ui.ViewHolder;

public class SelectableItemViewHolder<T> extends ViewHolder<T> {

    private final SingleSelectionManager singleSelectionManager;

    private final View expandableContainer;

    public SelectableItemViewHolder(ViewGroup parent, int layout, SingleSelectionManager singleSelectionManager) {
        super(parent, layout);
        this.singleSelectionManager = singleSelectionManager;

        prepareEventListener(itemView);

        expandableContainer = itemView.findViewById(R.id.details);
    }

    protected void prepareEventListener(View itemView) {
        itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                toggleSelection();
            }
        });
    }

    @Override
    protected void onBind(T item) {
        super.onBind(item);

        final boolean selected = isSelected();
        expandableContainer.setVisibility(selected ? View.VISIBLE : View.GONE);
    }

    protected boolean isSelected() {
        return singleSelectionManager.isSelected(getAdapterPosition());
    }

    public void toggleSelection() {
        final int position = getAdapterPosition();
        if (singleSelectionManager.isSelected(position)) {
            singleSelectionManager.clearSelection();
        } else {
            singleSelectionManager.setSelection(position);
        }
    }
}

File: modules/core/src/main/java/de/deutschebahn/bahnhoflive/view/SelectableItemViewHolder.kt
/*
 * SPDX-FileCopyrightText: 2020 DB Station&Service AG <bahnhoflive-opensource@deutschebahn.com>
 *
 * SPDX-License-Identifier: Apache-2.0
 */
package de.deutschebahn.bahnhoflive.view

import android.view.View
import androidx.core.view.isVisible
import de.deutschebahn.bahnhoflive.R
import de.deutschebahn.bahnhoflive.ui.ViewHolder

open class SelectableItemViewHolder<T>(
    view: View,
    private val singleSelectionManager: SingleSelectionManager?
) : ViewHolder<T>(view) {

    private val expandableContainer: View = itemView.findViewById(R.id.details)

    init {
        itemView.setOnClickListener { toggleSelection() }
    }


    override fun onBind(item: T?) {
        super.onBind(item)
        expandableContainer.isVisible = isSelected()
    }

    open fun isSelected(): Boolean {
        return singleSelectionManager?.isSelected(absoluteAdapterPosition) ?: true
    }

    fun toggleSelection() {
        if (singleSelectionManager == null) {
            if (expandableContainer.visibility == View.GONE) expandableContainer.visibility =
                View.VISIBLE else expandableContainer.visibility = View.GONE
        } else {
            val position = absoluteAdapterPosition
            if (singleSelectionManager.isSelected(position)) {
                singleSelectionManager.clearSelection()
            } else {
                singleSelectionManager.selection = position
            }
        }
    }
}
